plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/AbstractTestMySqlTableStatistics.java

```diff
@@ -0,0 +1,461 @@
+    {
+        assertEquals(columnNdvs.keySet(), columnNullFractions.keySet());
+        List<String> reportedColumns = stream(statsResult)
+                .map(row -> row.getField(0)) // column name
+                .filter(Objects::nonNull)
+                .map(String.class::cast)
+                .collect(toImmutableList());
+        assertThat(reportedColumns)
+                .containsOnlyOnce(columnNdvs.keySet().toArray(new String[0]));
+
+        double tableCardinality = getTableCardinalityFromStats(statsResult);
+        for (MaterializedRow row : statsResult) {
+            if (row.getField(0) == null) {
+                continue;
+            }
+            String columnName = (String) row.getField(0);
+            verify(columnNdvs.containsKey(columnName));
+            Integer expectedNdv = columnNdvs.get(columnName);
+            verify(columnNullFractions.containsKey(columnName));
+            Double expectedNullFraction = columnNullFractions.get(columnName);
+
+            Double dataSize = (Double) row.getField(1);
+            if (dataSize != null) {
+                assertThat(dataSize).as("Data size for " + columnName)
+                        .isEqualTo(0);
+            }
+
+            AbstractDoubleAssert<?> ndvAssertion = assertThat((Double) row.getField(2)).as("NDV for " + columnName);
+            if (expectedNdv == null) {
+                ndvAssertion.isNull();
+                assertNull(row.getField(3), "null fraction for " + columnName);
+            }
+            else {
+                ndvAssertion.isBetween(expectedNdv * 0.5, min(expectedNdv * 4.0, tableCardinality)); // [-50%, +300%] but no more than row count
+                assertThat((Double) row.getField(3)).as("Null fraction for " + columnName)
+                        .isBetween(expectedNullFraction * 0.4, min(expectedNullFraction * 1.1, 1.0));
+            }
+
+            assertNull(row.getField(4), "min");
+            assertNull(row.getField(5), "max");
+        }
+    }
+
+    protected static double getTableCardinalityFromStats(MaterializedResult statsResult)
+    {
+        MaterializedRow lastRow = statsResult.getMaterializedRows().get(statsResult.getRowCount() - 1);
+        assertNull(lastRow.getField(0));
+        assertNull(lastRow.getField(1));
+        assertNull(lastRow.getField(2));
+        assertNull(lastRow.getField(3));
+        assertNull(lastRow.getField(5));
+        assertNull(lastRow.getField(6));
+        assertEquals(lastRow.getFieldCount(), 7);
+        assertNotNull(lastRow.getField(4));
+        return ((Number) lastRow.getField(4)).doubleValue();
+    }
+
+    protected static class MapBuilder<K, V>
+    {
+        private final Map<K, V> map = new HashMap<>();
+
+        public MapBuilder<K, V> put(K key, V value)
+        {
+            checkArgument(!map.containsKey(key), "Key already present: %s", key);
```

Comment:
ImmutableMap.buildOrThrow does this already AFAIK.

Reply:
correct, but it doesn't allow null values